feat: env0 deployment logs

meniRoy opened the pull request:
when a user click on environment or redeploy (abort, cancel, approve etc) an environment we fetch the logs of the latest deployment  in this environment and show the logs in `env0 logs` output channel.
if the user clicks on different environment or redeploy a different environment we stop fetching the logs of the previous environment, we clear the output channel, and start to log the the latest deployment for the current selected environment.



https://github.com/env0/vscode-env0/assets/33559192/fa14c788-4dd6-466c-b707-cc2f7eb00bd4

meniRoy commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add new ENV0 LOGS tab](https://trello.com/c/4d5fEPFu/10034-add-new-env0-logs-tab)

meniRoy commented:
to avoid wasting time i think it will be better to add test in this PR or a different PR after the logic of this PR approved

sagilaufer1992 commented on src/api-client.ts:
```diff
@@ -59,22 +59,39 @@ class ApiClient {
+    deploymentLogId: string,
+    abortController?: AbortController
+  ) {
+    const res = await this.instance.get(
```
no need to shorten `response` to `res`. please be consistent across a file

meniRoy replied:
24dbf1ce4ada2bd10c97bad950028b9cb31553d2

sagilaufer1992 commented on src/environment-logs-provider.ts:
```diff
@@ -0,0 +1,168 @@
+      this.abortController
+    );
+
+    if (deployment.status === DeploymentStatus.QUEUED) {
+      this.log(`Deployment is queued! Waiting for it to start...`);
+    }
```
update `previousStatus`, you already queried the deployment, no need for another logging within the while if it's still `QUEUED`

sagilaufer1992 replied:
you didn't store the status, it won't reach the `previous === QUEUED`, no?

meniRoy replied:
i am not sure i understand,
but i did assign the status value to the previousStatus variable in the end of the loop [here](https://github.com/env0/vscode-env0/pull/40/files#diff-55ec115f8bce8531aada11dab10df383176a314d2a8491ee5cac38f2c6da6d81R95)

sagilaufer1992 replied:
you store _only_ at the end of the loop. you put a `continue` in the condition, so the value for `previousStatus` isn't updated

meniRoy replied:
Good catch, 16fbf35f896c503d459340da44d8f23f7a3e4860

sagilaufer1992 commented on src/environment-logs-provider.ts:
```diff
@@ -0,0 +1,168 @@
+      ) {
+        this.log(
+          `Deployment reached its turn! ${
+            type === "deploy" ? "Deploying" : "Destroying"
```
can it not be a drift, task?

meniRoy replied:
asked about that in vscode channel

sagilaufer1992 replied:
as I understand, we shouldn't log anything that is destroy/deploy/resumes. is that handled in the code? where do we filter out the other types?

meniRoy replied:
we logging the latest deployment for environment when selected or when environment action triggered by the user.
As discussed in the channel, I updated the log 6aeaac2179df3364981749bb923738786fa4eef2

sagilaufer1992 commented on src/environment-logs-provider.ts:
```diff
@@ -0,0 +1,168 @@
+    let startTime: number | string;
+
+    do {
+      const steps = await apiClient.getDeploymentSteps(
+        deploymentLogId,
+        this.abortController
+      );
+
+      const { status } = steps.find((step) => step.name === stepName) || {};
```
why not querying a single step? it seems we have such an endpoint

sagilaufer1992 replied:
https://github.com/env0/env0/blob/2ceeebb685dd92379d4ec874caabc0ffd6ce64ca/services/deployment-step-service/openapi.json#L162

meniRoy replied:
this endpoint return the log not the status

sagilaufer1992 commented on src/environment-logs-provider.ts:
```diff
@@ -0,0 +1,168 @@
+      this.log(`Deployment is queued! Waiting for it to start...`);
+    }
+
+    while (!this.isAborted) {
```
extract `while` content, and use `setInterval` instead of sleeping for delaying next execution

meniRoy replied:
I have concerns about using `setInterval`. In the current implementation with the `while` loop, I can use await to ensure that processDeploymentSteps completes before proceeding. This offers controlled management. However, with setInterval, there's uncertainty regarding how long processDeploymentSteps might take to finish. This could make synchronization and management more challenging, especially since other internal functions like writeDeploymentStepLog also wait for certain steps to complete.

sagilaufer1992 replied:
👍

sagilaufer1992 reviewed (commented):
left comments on code, please address

meniRoy commented:
> @meniRoy I see that the logs of "Active" environment are being fetched slowly, like an IN_PROGRESS environment. Isn't it possible to fetch all steps logs at once?

i didn't see any endpoint to fetch all steps logs at once

liranfarage89 commented:
> > @meniRoy I see that the logs of "Active" environment are being fetched slowly, like an IN_PROGRESS environment. Isn't it possible to fetch all steps logs at once?
> 
> i didn't see any endpoint to fetch all steps logs at once

but it looks like it takes more time than the frontend... ( for non IN-PROGRESS environments). Maybe the code "aggregates" the logs and then pipe it at once.
But there is not reason to wait a lot of time of logs that are already existing.

meniRoy commented:
> > > @meniRoy I see that the logs of "Active" environment are being fetched slowly, like an IN_PROGRESS environment. Isn't it possible to fetch all steps logs at once?
> > 
> > 
> > i didn't see any endpoint to fetch all steps logs at once
> 
> but it looks like it takes more time than the frontend... ( for non IN-PROGRESS environments). Maybe the code "aggregates" the logs and then pipe it at once. But there is not reason to wait a lot of time of logs that are already existing.

[56bb679](https://github.com/env0/vscode-env0/pull/40/commits/56bb679a3151c6d0032b47ab6373da6a6c8bc5ee)

meniRoy commented:
I found a bug, when the user clicks on redeploy we log the previous deployment instead of logging the deployment triggered by the user action.
I will fix that after I merge this [PR](https://github.com/env0/vscode-env0/pull/46)

meniRoy commented:
> I found a bug, when the user clicks on redeploy we log the previous deployment instead of logging the deployment triggered by the user action. I will fix that after I merge this [PR](https://github.com/env0/vscode-env0/pull/46)

solved f6f864e45d01b842eaafc95f271184d9270e8462
Added one more commit ["don't restart logs if no environment selected"](https://github.com/env0/vscode-env0/pull/40/commits/601d517136676ec0f51d7c45bb32bdd5a8b17ab5)
Updated the video in the PR description

sagilaufer1992 commented:
while playing with the extension on my machine, I saw that when I hit approve/redeploy, the selected env isn't changing, but the logs are for the redeployed/approved env. @meniRoy please fix this issue

meniRoy commented:
> while playing with the extension on my machine, I saw that when I hit approve/redeploy, the selected env isn't changing, but the logs are for the redeployed/approved env. @meniRoy please fix this issue

solved 3d575729b64fe79251f86c9fda8fa2d570581be5

also committed minor fix to set max listeners of the abortController to prevent NodeJs warnings for more then 10 listeners c9e2170c04c82d4537b446c8d62c831100f03920

sagilaufer1992 reviewed (approved):
LGTM, the changes to the issues raised seem ok, QA passed, and other additions are explained.
just remove or explain what is the commented addition please

meniRoy commented on src/env0-environments-provider.ts:
```diff
@@ -62,6 +62,10 @@ export class Env0EnvironmentsProvider
     );
   }
 
+  getParent(element: Environment): vscode.ProviderResult<Environment> {
+    return null;
+  }
+
```
Because I use `treeview.reveal` I need to implement this function, since we don't have any nested elements, I always return now
